### DISCUSSION TITLE
Use of parameters in chunk options

### DISCUSSION THREAD

**Discussion by Dobrokhotov1989** [quarto-dev/quarto-cli#1596] [Q&A] [answered]
Hi there!

I'm trying to migrate from RMarkdown to `quarto`. I'm lucky because I've heard about `quatro` just a few days after I began a new project.
In RMarkdown, I used params to conditionally execute some chunks of code (setting params in yaml and then passing them into chunk option 'eval').  I've tried to do this trick in `quarto`, but the chunk option is not recognized as a parameter, but just as a character string and thus just ignored. Is it possible to achieve such behavior in `quarto`?

Example:
In RMarkdown (works): 
````
```{r eval=params$ChunkNumberOne}
SomeCode <- That + ShouldNotBe + Evaluated
```
````
In quarto (doesn't work:
````
```{r}
#| eval: params$ChunkNumberOne

SomeCode <- That + ShouldNotBe + Evaluated
```
````

**Comment by mcanouil** [comment, marked as answer]
you have to use the keyword/command `!expr` as noted in https://quarto.org/docs/computations/r.html#chunk-options

```{r}
#| eval: !expr params$ChunkNumberOne

SomeCode <- That + ShouldNotBe + Evaluated
```

**Reply by Dobrokhotov1989** [reply]
Sorry, I've totally missed this point in docs. Thanks a lot!

**Reply by Dobrokhotov1989** [reply]
BTW, is it possible to write params in a separate file? When I stated params directly in a file YAML header - everything works. But when I tried to write them _quarto.yml (and used `quarto render` ) or other .yml file (`quarto render index.qmd --execute-params params.yml`) - params are not recognized, and I get an error. 

```
Error: object 'params' not found
Error in yaml::yaml.load(yaml, eval.expr = TRUE) :
  Could not evaluate expression: params$chunkEval
Calls: .main ... parse_block -> partition_yaml_options -> <Anonymous>
Execution halted
```

**Reply by mcanouil** [reply]
I don't know especially since I don't know how you wrote the different files.
From the docs, I don't think it is supposed to work in `_quarto.yml` but it is supposed to for the other method (https://quarto.org/docs/computations/parameters.html).
Please provide a reproducible example.

**Reply by hebertodelrio** [reply]
Don’t you think that this can be treated similarly like the inline code case (https://github.com/quarto-dev/quarto-cli/discussions/1520#discussioncomment-3291022)?

extract the top comments (#|) create a fake code cell with them and execute it therefore every occurrence of {{ code }} will be replaced with the appropriate result, and replace the top comments with the result.

I can tinkle with the idea and see how far I can get. One issue I have is to understand the definition of options (#|) do they must be at the very top of the code cell and no empty newlines in between them, first empty newline determines the end of the options? Or there is something different?

Again if this work will be as a Notebook filter

**Reply by mcanouil** [reply]
Parameters are already handled in Quarto especially using an external yaml file passed as parameter in Quarto CLI (at least according to documentation), thus the issue here seems to be a syntax error which we can't see without the actual code.

**Reply by hebertodelrio** [reply]
What I meant is that the use of !expr is only valid if you are using knitr, so people that uses Python or Julia (like me) are in a bit of disadvantage

Mote than parameters I was referring to chunk options with pseudo dynamic content

**Reply by mcanouil** [reply]
For Jupyter engine hence Python: https://quarto.org/docs/computations/parameters.html#jupyter
So you could get the same behaviour in python cells by adding if else statement rather than using the `eval` yaml cell option.

**Reply by Dobrokhotov1989** [reply]
@mcanouil, thank you for your help.

I have three files in the same project folder:

File `test.qmd`
````
## Test

```{r}
#| eval: !expr params$chunkeval
1 + 1
```

````

File `_quarto.yml`
````
project:
  title: "Test"

editor: visual
````

File `params.yml`

````
params:
  chunkeval: false
````

When I use command `quarto render test.qmd --execute-params params.yml` in terminal I got the following error:
```
processing file: test.qmd
Error: object 'params' not found
Error in yaml::yaml.load(yaml, eval.expr = TRUE) :
  Could not evaluate expression: params$chunkeval
Calls: .main ... parse_block -> partition_yaml_options -> <Anonymous>
Execution halted
```
